cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramNotificationSubscriberService.java

```diff
@@ -280,6 +281,18 @@ protected void storeMessageId(StructuredTableContext context, String messageId)
     getAppMetadataStore(context).persistSubscriberState(getTopicId().getTopic(), "", messageId);
   }
 
+  @Override
+  protected boolean shouldRunInSeparateTx(ImmutablePair<String, Notification> message) {
+    String clusterStatusStr = message.getSecond().getProperties().get(ProgramOptionConstants.CLUSTER_STATUS);
+    if (clusterStatusStr != null &&
+      ProgramRunClusterStatus.valueOf(clusterStatusStr) == ProgramRunClusterStatus.PROVISIONING){
+      // PROVISIONING notifications result in increment in programCountsTable which may result in DB deadlock
```

Comment:
would be good to also check that the topic is partitioned, as we can't deadlock with just the one partition.

Reply:
though if we're doing this, is there a need for the tx size?

Reply:
I put this behind a flag. Reason: hope is that realistically when number of pipelines is high, this does not happen. If it happens, then we can reduce txSize to avoid deadlocks. But just in case we start hitting it under very high load with very few pipelines, then we have the option to make this operation into separate txn.